# CGD Transcript Effects

The python modules in this directory are used to gather transcript information with proper HGVS nomenclature for import into CGD. 

The four main modules are ``tx_eff_control.py``, ``tx_eff_annovar.py``, ``tx_eff_hgvs.py``, and ``tx_eff_vcf.py``. Each module has a main function and is able to take command line parameters indicating the file(s) to read from and write to. Unless you are debugging or developing this project then you should only run ``tx_eff_control.py``, and let it call the other modules. 

## Configuration 

### Install Annovar

### Install HGVS


## Annovar

Before running Annovar you must first convert your variant list to a format Annovar understands. It is necessary to include the ``--includeinfo`` parameter so that all the information from the original VCF is included in the Annovar input file. 

```
./convert2annovar.pl --includeinfo \
                     --format vcf4old \
                     input.vcf > annovar.avinput
```

Run Annovar with the following parameters to add RefSeq transcripts and identify splice variants. 

```
./annotate_variation.pl -out annovar \
                        -build hg19 \
                        annovar.avinput \
                        ./humandb/ \
                        --transcript_function \
                        --hgvs \
                        --separate \
                        --dbtype refGeneWithVer \
                        --otherinfo \
                        --splicing_threshold 5 \
                        --exonicsplicing
```

The ``annotate_variation.pl`` script generates two output files: 
* annovar.exonic_variant_function
* annovar.variant_function

## The Transcript Effects Modules


### tx_eff_control.py 

The ``tx_eff_control.py`` script is the interface for calling the other five modules. This script takes as parameters an input VCF file, the variant_function and exonic_variant_function files generated by ``annotate_variation.pl``, and the name of a VCF file to write the results. 

The optional parameter ``--require_match`` may be used to indicate that the final list should only include transcripts that are common to both of the HGVS/UTA and Annovar sources. 

Example of how to run ``tx_eff_control.py``:

```
python src/edu/ohsu/compbio/txeff/tx_eff_control.py --in_vcf input.vcf \
                                                    --out_vcf cgd_tx_eff.vcf \
                                                    annovar.variant_function \
                                                    annovar.exonic_variant_function
```

### tx_eff_annovar.py

The ``tx_eff_annovar.py`` script reads the ``*.exonic_variant_function`` and ``*.variant_function`` files generated by Annovar and  merges the information from the two files into one record for each transcript. 

```
python src/edu/ohsu/compbio/txeff/tx_eff_annovar.py --out_file tx_eff_annovar_out.csv \
                                                    annovar.variant_function \
                                                    annovar.exonic_variant_function
```

### tx_eff_hgvs.py

The ``tx_eff_hgvs.py`` script takes as input the transcripts generated by ``tx_eff_annovar.py`` and fixes the nomenclature using the HGVS python library. 

Each variant in the input file is searched for in the HGVS UTA database. Next, each HGVS variant and transcript is merged with an Annovar record having a matching variant genotype and transcript. Not all of the transcripts returned by the HGVS python library will be matched with one of the Annovar transcripts. A common reason for mismatch is that RefSeq transcript versions are different. By default the final list of transcripts will include the transcripts that are merged from the Annovar and HGVS python library sources, as well as those that are unique to one or the other sources. The optional parameter ``--require_match`` may be used to indicate that the final list should only include transcripts that are common to both sources. 

```
python src/edu/ohsu/compbio/txeff/tx_eff_hgvs.py --in_file=tx_eff_annovar_out.csv \
                                                 --out_file=tx_eff_hgvs_out.csv \
                                                 --require_match
```

### tx_eff_ccds.py

The ``tx_eff_vcf.py`` script takes the transcripts generated by ``tx_eff_hgvs.py`` and , looks up each RefSeq id in the NCBI CCDS map, and creates a copy of the RefSeq transcript. 

### tx_eff_vcf.py

The ``tx_eff_vcf.py`` script takes as input the transcript list generated by ``tx_eff_hgvs.py`` and ``tx_eff_ccds.py``, and writes them out to a VCF file with the transcripts nomenclature placed in the INFO field. 

```
python src/edu/ohsu/compbio/txeff/tx_eff_vcf.py --in_vcf input.vcf \
                                                --in_csv tx_eff_hgvs_out.csv \
                                                --out_vcf cgd_tx_eff_vcf.vcf
```

The transcript-effects attributes in the INFO field are parallel arrays. For example the following attributes are populated for the variant 5-231111-T-C:

* TFX_BASE_POSITION=891:891:747:891:891;
* TFX_EXON=7::6:7:;
* TFX_GENE=SDHA::SDHA:SDHA:;
* TFX_HGVSC=c.891T>C:c.891T>C:c.747T>C:c.891T>C:c.891T>C;
* TFX_HGVSP1=p.(P297=):p.(P297=):p.(P249=):p.(P297=):p.(P297=);
* TFX_HGVSP3=p.(Pro297=):p.(Pro297=):p.(Pro249=):p.(Pro297=):p.(Pro297=);
* TFX_SPLICE=splicing::splicing:splicing:;
* TFX_TRANSCRIPT=NM_001330758.1:NM_004168.3:NM_001294332.1:NM_004168.4:NM_004168.2;
* TFX_VARIANT_EFFECT=synonymous SNV::synonymous SNV:synonymous SNV:;
* TFX_VARIANT_TYPE=exonic::exonic:exonic:;
* TFX_PROTEIN_TRANSCRIPT=NP_001317687.1:NP_004159.2:NP_001281261.1:NP_004159.2:NP_004159.2
* TFX_AMINO_ACID_POSITION=::splicing:;

## Additional Utilities

### avinput2vcf.py

### csv2avinput.py

### compare_cgd_tfx.py

### refseq_to_ccds.py

